quil-py/quil/program/__init__.pyi

```diff
@@ -175,6 +175,82 @@ class Program:
         while ensuring that unique value is not already in use within the program.
         """
         ...
+    def control_flow_graph(self) -> "ControlFlowGraph":
+        """
+        Return the `control flow graph`_ of the program.
+
+        .. _control flow graph: https://en.wikipedia.org/wiki/Control-flow_graph
+        """
+
+@final
+class BasicBlock:
+    def as_fixed_schedule(self, program: Program, include_zero_duration_instructions: bool) -> FixedSchedule:
+        """
+        Return the ``FixedSchedule`` representing the timing of the instructions within the block.
+
+        This schedule is computed by:
+
+        * Expanding each instruction within the block using the program's calibration definitions
+        * Resolving the `FixedSchedule` of the expanded instructions
+        * Mapping calibrated instructions back to the original instructions within this block, such that the
+          block's instruction is represented as a timespan encompassing all of its expanded instructions
+
+        If `include_zero_duration_instructions` is `True`, then the schedule will include instructions with zero duration
+        (such as `FENCE`). This may cause certain instructions to appear as unexpectedly long in duration.
+
+        :param program: The program containing the calibrations to be used to schedule this block. Generally,
+            this should be the program from which the block was extracted
+        :param include_zero_duration_instructions: Whether to include instructions with zero duration in the schedule
+
+        The following example demonstrates construction of such a schedule for a simple program without explicit control
+        flow (and thus with only one basic block):
+
+        .. example-code::
```

Comment:
This doesn't seem like an appropriate example; it's for `pyquil`, but `quil-py` should be able to be demonstrated stand-alone. An alternative might be something like:

```
from quil.program import Program

program = Program.parse("""
<some example instructions>
""")

print(program.to_quil())

control_flow_graph = program.control_flow_graph()
assert control_flow_graph.has_dynamic_control_flow() == False

basic_blocks = control_flow_graph.basic_blocks()
assert len(basic_blocks) == 1

schedule = blocks[0].as_fixed_schedule(program, False)
print(f"Duration = {schedule.duration()}")

print(schedule.items())
```

Reply:
Update: Using some appropriate calibration and frame definitions, and a short test case:
```
ISWAP 23 30
RX(pi) 23
RX(pi) 24
DELAY 23 24 2.48e-7
RX(pi) 24
```

Then bypassing the `__repr__()` comment the following follow-on code can create some reasonable output:
```
control_flow_graph = program.control_flow_graph()
assert control_flow_graph.has_dynamic_control_flow() == False

basic_blocks = control_flow_graph.basic_blocks()
assert len(basic_blocks) == 1

basic_block = basic_blocks[0]

instructions = basic_block.instructions()

schedule = basic_block.as_fixed_schedule(program, False)

print(f"Duration = {schedule.duration()}")

for schedule_item in schedule.items():
    start = schedule_item.time_span.start
    duration = schedule_item.time_span.duration
    end = start + duration

    instruction = instructions[schedule_item.instruction_index]

    print(f"{instruction} => {start} to {end}")
```

The output of the demonstration:
```
Gate(Gate { name: "RX", parameters: [PiConstant], qubits: [Fixed(24)], modifiers: [] }) => 0.0 to 4e-08
Gate(Gate { name: "RX", parameters: [PiConstant], qubits: [Fixed(23)], modifiers: [] }) => 5.6000000000000005e-08 to 9.600000000000001e-08
Gate(Gate { name: "ISWAP", parameters: [], qubits: [Fixed(23), Fixed(30)], modifiers: [] }) => 0.0 to 5.6000000000000005e-08
Gate(Gate { name: "RX", parameters: [PiConstant], qubits: [Fixed(24)], modifiers: [] }) => 3.44e-07 to 3.84e-07
Delay(Delay { duration: Number(Complex { re: 2.48e-7, im: 0.0 }), frame_names: [], qubits: [Fixed(23), Fixed(24)] }) => 9.600000000000001e-08 to 3.44e-07
```

Observations:
- Seeing `Gate` twice in the nested representations seems unnecessary?
- Floating point precision isn't ideal for a nanosecond-scale system; comment captured elsewhere.
- A complex numeric value doesn't seem to be the right type for the delay?

Reply:
> Seeing Gate twice in the nested representations seems unnecessary?

Yeah, although this is not new to this PR - it's the `Debug` output from the Rust side.

> A complex numeric value doesn't seem to be the right type for the delay?

It's because it's an expression, and all numbers in `Expression` are represented as `Complex64`

Reply:
> This doesn't seem like an appropriate example; it's for pyquil, but quil-py should be able to be demonstrated stand-alone

Of course! Can't believe I missed that. Thanks

Reply:
Updated example looks good.